File: C++/044. Wildcard Matching.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    bool isMatch(std::string s, std::string p) {
        int len_p = int(p.size()), len_s = int(s.size());
        std::vector<std::vector<bool>> dp(p.size() + 1, std::vector<bool>(s.size() + 1, false));
        dp[0][0] = true;
        s = "#" + s;
        p = "#" + p;
        for (int i = 1; i <= len_p && p[i] == '*'; i++)
            dp[i][0] = true;
        for (int i = 1; i <= len_p; i++) {
            for (int j = 1; j <= len_s; j++) {
                if (dp[i - 1][j - 1] && (p[i] == '?' || p[i] == s[j]))
                    dp[i][j] = true;
                else if (p[i] == '*') {
                    for (int k = j; k >= 0; k--) {
                        if (dp[i - 1][k] == true) {
                            dp[i][j] = true;
                            break;
                        }
                    }
                }
            }
        }
        return dp[len_p][len_s];
    }
};

File: Go/044. Wildcard Matching.go
package main

func isMatch(s string, p string) bool {
	lenP, lenS := len(p), len(s)
	dp := make([][]bool, lenP + 1, lenP + 1)
	for i := 0; i <= lenP; i++ {
		dp[i] = make([]bool, lenS + 1, lenS + 1)
	}
	s, p, dp[0][0] = "#" + s, "#" + p, true
	for i := 1; i <= lenP && p[i] == '*' ; i++ {
		dp[i][0] = true
	}
	for i := 1; i <= lenP; i++ {
		for j := 1; j <= lenS; j++ {
			if (p[i] == '?' || p[i] == s[j]) && dp[i - 1][j - 1] {
				dp[i][j] = true
			} else if p[i] == '*' {
				for k := j; k >= 0; k-- {
					if dp[i - 1][k] {
						dp[i][j] = true
						break
					}
				}
			}
		}
	}
	return dp[lenP][lenS]
}